// resilient-llm/dist/lib/ProviderRegistry.js
/**
 * Provider Registry - Centralized configuration and model management.
 * Singleton pattern for global provider configuration.
 * Handles provider configuration, API key management, and model fetching/caching.
 */
class ProviderRegistry {
    static #initialized = false;
    static #providers = new Map();
    /** Map<providerName, Map<modelName, UnifiedModel>> */
    static #modelsCache = new Map();
    /** Runtime API keys (stored separately from config to avoid serialization). Map<providerName, apiKey> */
    static #apiKeys = new Map();
    /** Default provider configurations (openai, anthropic, google, ollama). */
    static DEFAULT_PROVIDERS = {
        openai: {
            name: 'openai',
            displayName: 'OpenAI',
            chatApiUrl: 'https://api.openai.com/v1/chat/completions',
            modelsApiUrl: 'https://api.openai.com/v1/models',
            docsUrl: "https://developers.openai.com/api/reference/resources/chat/subresources/completions/methods/create",
            envVarNames: ['OPENAI_API_KEY'],
            defaultModel: 'gpt-5-nano',
            apiVersion: null,
            iconUrl: null,
            customHeaders: {},
            authConfig: {
                type: 'header',
                headerName: 'Authorization',
                headerFormat: 'Bearer {key}'
            },
            parseConfig: {
                modelsPath: 'data',
                idField: 'id',
                nameField: 'id',
                displayNameField: null,
                contextWindowField: null,
                idPrefix: null
            },
            chatConfig: {
                messageFormat: 'openai',
                responseParsePath: 'choices[0].message.content',
                toolSchemaType: 'openai',
                structuredOutputRequestField: 'response_format',
            },
            active: true
        },
        anthropic: {
            name: 'anthropic',
            displayName: 'Anthropic',
            chatApiUrl: 'https://api.anthropic.com/v1/messages',
            modelsApiUrl: 'https://api.anthropic.com/v1/models',
            docsUrl: "https://platform.claude.com/docs/en/api/messages/create",
            envVarNames: ['ANTHROPIC_API_KEY'],
            defaultModel: 'claude-haiku-4-5-20251001',
            apiVersion: '2023-06-01',
            iconUrl: null,
            customHeaders: {
                'anthropic-version': '2023-06-01'
            },
            authConfig: {
                type: 'header',
                headerName: 'x-api-key',
                headerFormat: '{key}'
            },
            parseConfig: {
                modelsPath: 'data',
                idField: 'id',
                nameField: 'id',
                displayNameField: 'display_name',
                contextWindowField: null,
                idPrefix: null
            },
            chatConfig: {
                messageFormat: 'anthropic',
                responseParsePath: 'content[0].text',
                toolSchemaType: 'anthropic',
                structuredOutputRequestField: 'output_config',
            },
            active: true
        },
        google: {
            name: 'google',
            displayName: 'Google',
            chatApiUrl: 'https://generativelanguage.googleapis.com/v1beta/openai/chat/completions',
            modelsApiUrl: 'https://generativelanguage.googleapis.com/v1/models',
            docsUrl: "https://ai.google.dev/api",
            envVarNames: ['GEMINI_API_KEY', 'GOOGLE_API_KEY', 'GOOGLE_GENERATIVE_AI_API_KEY'],
            defaultModel: 'gemini-2.0-flash',
            apiVersion: null,
            iconUrl: null,
            customHeaders: {},
            authConfig: {
                type: 'query',
                queryParam: 'key'
            },
            endpointAuthConfigs: {
                '/chat/completions': {
                    type: 'header',
                    headerName: 'Authorization',
                    headerFormat: 'Bearer {key}'
                },
                '/models': {
                    type: 'query',
                    queryParam: 'key'
                }
            },
            parseConfig: {
                modelsPath: 'models',
                idField: 'name',
                nameField: 'name',
                displayNameField: 'displayName',
                contextWindowField: 'inputTokenLimit',
                idPrefix: 'models/'
            },
            chatConfig: {
                messageFormat: 'openai',
                responseParsePath: 'choices[0].message.content',
                toolSchemaType: 'openai',
                structuredOutputRequestField: 'response_format',
            },
            active: true
        },
        ollama: {
            name: 'ollama',
            displayName: 'Ollama',
            chatApiUrl: process.env?.OLLAMA_API_URL || 'http://localhost:11434/api/generate',
            modelsApiUrl: `${process.env?.OLLAMA_API_URL || 'http://localhost:11434'}/api/tags`,
            envVarNames: ['OLLAMA_API_KEY'],
            defaultModel: 'llama3.1:8b',
            apiVersion: null,
            iconUrl: null,
            customHeaders: {},
            authConfig: {
                type: 'header',
                headerName: 'Authorization',
                headerFormat: 'Bearer {key}',
                optional: true
            },
            parseConfig: {
                modelsPath: 'models',
                idField: 'name',
                nameField: 'name',
                displayNameField: null,
                contextWindowField: null,
                idPrefix: null
            },
            chatConfig: {
                messageFormat: 'openai',
                responseParsePath: 'response',
                toolSchemaType: 'openai',
                structuredOutputRequestField: 'response_format',
            },
            active: true
        }
    };
    /**
     * Normalize provider name (lowercase, no spaces) for case-insensitive matching.
     * @param providerName - Provider identifier (may have mixed case/spaces)
     * @returns Normalized provider name
     * @private
     */
    static #normalizeProviderName(providerName) {
        if (!providerName || typeof providerName !== 'string') {
            return '';
        }
        return providerName.replace(/\s+/g, '').toLowerCase();
    }
    /** Initialize registry with default providers. Auto-called on first use; can be called explicitly. */
    static init() {
        if (this.#initialized) {
            return;
        }
        for (const [name, config] of Object.entries(this.DEFAULT_PROVIDERS)) {
            this.#providers.set(name, { ...config });
        }
        this.#initialized = true;
    }
    /**
     * Configure or update a provider. Uses merge strategy: new config merges with existing.
     * Supports baseUrl (for Ollama/OpenAI-compatible), chatApiUrl, modelsApiUrl, envVarNames,
     * apiKey, defaultModel, authConfig, parseConfig, chatConfig, active. See ConfigureInput.
     * @param providerName - Provider identifier
     * @param config - Provider configuration (partial merge)
     * @returns The merged provider config
     */
    static configure(providerName, config) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const existing = this.#providers.get(providerName) || {};
        let chatApiUrl = config.chatApiUrl || existing.chatApiUrl;
        let modelsApiUrl = config.modelsApiUrl || existing.modelsApiUrl;
        if (config.baseUrl) {
            const baseUrl = config.baseUrl.replace(/\/$/, '');
            if (providerName === 'ollama') {
                if (!existing.chatApiUrl && !config.chatApiUrl)
                    chatApiUrl = `${baseUrl}/api/generate`;
                if (!existing.modelsApiUrl && !config.modelsApiUrl)
                    modelsApiUrl = `${baseUrl}/api/tags`;
            }
            else {
                if (!existing.chatApiUrl && !config.chatApiUrl)
                    chatApiUrl = `${baseUrl}/v1/chat/completions`;
                if (!existing.modelsApiUrl && !config.modelsApiUrl)
                    modelsApiUrl = `${baseUrl}/v1/models`;
            }
        }
        if (!chatApiUrl && !existing.chatApiUrl) {
            console.warn(`Please set the chatApiUrl (or baseUrl) for provider "${providerName}". This provider won't work without it.`);
        }
        let apiKeyToStore = null;
        if (config.apiKey !== undefined) {
            apiKeyToStore = config.apiKey;
            delete config.apiKey;
        }
        const merged = {
            name: providerName,
            displayName: config.displayName || existing.displayName || providerName,
            chatApiUrl: chatApiUrl || '',
            modelsApiUrl: modelsApiUrl,
            envVarNames: config.envVarNames || existing.envVarNames || [],
            defaultModel: config.defaultModel || existing.defaultModel || '',
            apiVersion: config.apiVersion !== undefined ? config.apiVersion : (existing.apiVersion ?? null),
            iconUrl: config.iconUrl || existing.iconUrl || null,
            customHeaders: { ...(existing.customHeaders || {}), ...(config.customHeaders || {}) },
            authConfig: config.authConfig
                ? { ...(existing.authConfig || {}), ...config.authConfig }
                : (existing.authConfig || {
                    type: 'header',
                    headerName: 'Authorization',
                    headerFormat: 'Bearer {key}'
                }),
            parseConfig: config.parseConfig
                ? { ...(existing.parseConfig || {}), ...config.parseConfig }
                : (existing.parseConfig || {
                    modelsPath: 'data',
                    idField: 'id',
                    nameField: 'id',
                    displayNameField: null,
                    contextWindowField: null,
                    idPrefix: null
                }),
            chatConfig: config.chatConfig
                ? { ...(existing.chatConfig || {}), ...config.chatConfig }
                : (existing.chatConfig || {
                    messageFormat: 'openai',
                    responseParsePath: 'choices[0].message.content',
                    toolSchemaType: 'openai',
                    structuredOutputRequestField: 'response_format',
                }),
            active: config.active !== undefined ? config.active : (existing.active !== undefined ? existing.active : true)
        };
        this.#providers.set(providerName, merged);
        if (apiKeyToStore) {
            this.#apiKeys.set(providerName, apiKeyToStore);
        }
        this.#modelsCache.delete(providerName);
        return { ...merged };
    }
    /** Get provider config by name (null if not found). */
    static get(providerName) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.#providers.get(providerName);
        return provider ? { ...provider } : null;
    }
    /** List all providers, optionally filtered by options.active. */
    static list(options = {}) {
        this.init();
        let providers = Array.from(this.#providers.values());
        if (options.active !== undefined) {
            providers = providers.filter(p => p.active === options.active);
        }
        return providers.map(p => ({ ...p }));
    }
    static #getApiKey(providerName) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.get(providerName);
        if (!provider) {
            return null;
        }
        const storedKey = this.#apiKeys.get(providerName);
        if (storedKey) {
            return storedKey;
        }
        if (provider.envVarNames && provider.envVarNames.length > 0) {
            if (typeof process !== 'undefined' && process.env) {
                for (const envVarName of provider.envVarNames) {
                    if (process.env[envVarName]) {
                        return process.env[envVarName];
                    }
                }
            }
        }
        const authConfig = provider.authConfig || {};
        if (authConfig.optional) {
            return null;
        }
        return null;
    }
    /** Get the chat API URL for a provider (null if not configured). */
    static getChatApiUrl(providerName) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.get(providerName);
        return provider ? provider.chatApiUrl : null;
    }
    /** Get auth headers (and customHeaders) for a provider; uses endpoint-specific auth when endpointAuthConfigs match the URL. */
    static getHeaders(providerName, defaultHeaders = {}) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.get(providerName);
        if (!provider) {
            return defaultHeaders;
        }
        return { ...defaultHeaders, ...provider.customHeaders };
    }
    /** Get chat config (messageFormat, responseParsePath, toolSchemaType) for a provider. */
    static getChatConfig(providerName) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.get(providerName);
        return provider?.chatConfig || null;
    }
    /** Whether an API key is set for the provider (from config or environment). */
    static hasApiKey(providerName) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        return this.#getApiKey(providerName) !== null;
    }
    /**
     * Build headers with auth (and customHeaders). Uses endpoint-specific auth from endpointAuthConfigs when apiUrl matches.
     * @param providerName - Provider identifier
     * @param apiKey - Optional API key (otherwise from registry/env)
     * @param defaultHeaders - Base headers to merge
     * @param apiUrl - Optional URL to select endpoint-specific auth (e.g. for query vs header auth per endpoint)
     */
    static buildAuthHeaders(providerName, apiKey = null, defaultHeaders = {}, apiUrl = null) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.get(providerName);
        if (!provider) {
            return defaultHeaders;
        }
        if (!apiKey) {
            apiKey = this.#getApiKey(providerName);
        }
        let authConfig = provider.authConfig || {
            type: 'header',
            headerName: 'Authorization',
            headerFormat: 'Bearer {key}'
        };
        if (apiUrl && provider.endpointAuthConfigs) {
            const matchingPattern = Object.keys(provider.endpointAuthConfigs)
                .filter(pattern => apiUrl.includes(pattern))
                .sort((a, b) => b.length - a.length)[0];
            if (matchingPattern) {
                authConfig = provider.endpointAuthConfigs[matchingPattern];
            }
        }
        const headers = { ...defaultHeaders };
        if (apiKey && authConfig.type === 'header') {
            const headerName = authConfig.headerName || 'Authorization';
            const headerFormat = authConfig.headerFormat || 'Bearer {key}';
            const headerValue = headerFormat.replace('{key}', apiKey);
            headers[headerName] = headerValue;
        }
        return { ...headers, ...provider.customHeaders };
    }
    /** Build final API URL; adds query-param auth when authConfig.type is 'query'. Auto-detects endpoint-specific auth from endpointAuthConfigs. */
    static buildApiUrl(providerName, apiUrl, apiKey = null) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const provider = this.get(providerName);
        if (!provider) {
            return apiUrl;
        }
        let authConfig = provider.authConfig || {};
        if (provider.endpointAuthConfigs) {
            const matchingPattern = Object.keys(provider.endpointAuthConfigs)
                .filter(pattern => apiUrl.includes(pattern))
                .sort((a, b) => b.length - a.length)[0];
            if (matchingPattern) {
                authConfig = provider.endpointAuthConfigs[matchingPattern];
            }
        }
        if (authConfig.type === 'query') {
            if (!apiKey) {
                apiKey = this.#getApiKey(providerName);
            }
            if (apiKey) {
                const separator = apiUrl.includes('?') ? '&' : '?';
                const paramName = authConfig.queryParam || 'key';
                return `${apiUrl}${separator}${paramName}=${encodeURIComponent(apiKey)}`;
            }
        }
        return apiUrl;
    }
    /** Fetch and cache models for a provider (or return all cached if providerName is null). */
    static async getModels(providerName = null, apiKey = null) {
        this.init();
        if (providerName) {
            providerName = this.#normalizeProviderName(providerName);
            const cached = this.#modelsCache.get(providerName);
            if (cached && cached.size > 0) {
                return Array.from(cached.values());
            }
            const provider = this.get(providerName);
            if (!provider) {
                return [];
            }
            if (!provider.modelsApiUrl) {
                return [];
            }
            try {
                const keyToUse = apiKey || this.#getApiKey(providerName);
                const authConfig = provider.authConfig || {};
                if (!keyToUse && !authConfig.optional) {
                    console.warn(`API key not available for ${providerName}. Neither provided nor available in environment variables.`);
                }
                const models = await this.#fetchModels(providerName, provider, keyToUse);
                const modelMap = new Map();
                models.forEach(model => {
                    modelMap.set(model.id, model);
                });
                this.#modelsCache.set(providerName, modelMap);
                return models;
            }
            catch (error) {
                console.error(`Error fetching models for ${providerName}:`, error);
                return [];
            }
        }
        else {
            const allModels = [];
            for (const [, modelMap] of this.#modelsCache.entries()) {
                allModels.push(...Array.from(modelMap.values()));
            }
            return allModels;
        }
    }
    /** Get a single model by provider and model name (fetches models if not cached). */
    static async getModel(providerName, modelName, apiKey = null) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        const cached = this.#modelsCache.get(providerName);
        if (cached && cached.has(modelName)) {
            return cached.get(modelName);
        }
        const models = await this.getModels(providerName, apiKey);
        return models.find(m => m.id === modelName) || null;
    }
    static saveModel(providerName, modelName, modelData) {
        this.init();
        providerName = this.#normalizeProviderName(providerName);
        if (!this.#modelsCache.has(providerName)) {
            this.#modelsCache.set(providerName, new Map());
        }
        const modelMap = this.#modelsCache.get(providerName);
        modelMap.set(modelName, {
            ...modelData,
            id: modelName,
            provider: providerName
        });
    }
    static async #fetchModels(providerName, provider, apiKey) {
        let url = provider.modelsApiUrl;
        const defaultHeaders = { 'Content-Type': 'application/json' };
        if (!url)
            return [];
        url = this.buildApiUrl(providerName, url, apiKey);
        let headers = this.buildAuthHeaders(providerName, apiKey, defaultHeaders, url);
        if (provider.apiVersion && providerName === 'anthropic' && !headers['anthropic-version']) {
            headers['anthropic-version'] = provider.apiVersion;
        }
        const response = await fetch(url, {
            method: 'GET',
            headers
        });
        if (!response.ok) {
            throw new Error(`${providerName} API error: ${response.status} ${response.statusText}`);
        }
        const data = await response.json();
        return this.#parseModels(providerName, data);
    }
    static #parseModels(providerName, data) {
        const provider = this.get(providerName);
        const parseConfig = provider?.parseConfig || {
            modelsPath: 'data',
            idField: 'id',
            nameField: 'id',
            displayNameField: null,
            contextWindowField: null,
            idPrefix: null
        };
        let models = [];
        if (parseConfig.modelsPath) {
            const pathParts = parseConfig.modelsPath.split('.');
            let current = data;
            for (const part of pathParts) {
                if (current && typeof current === 'object' && part in current) {
                    current = current[part];
                }
                else {
                    current = null;
                    break;
                }
            }
            if (Array.isArray(current)) {
                models = current;
            }
        }
        if (models.length === 0) {
            const d = data;
            models = (d.data || d.models || d.items || d.results || []);
        }
        if (!Array.isArray(models)) {
            console.warn(`Unexpected response format for ${providerName}. Expected array of models.`);
            return [];
        }
        return models.map(model => {
            if (!model || typeof model !== 'object') {
                return null;
            }
            const m = model;
            let id = (this.#getNestedValue(m, parseConfig.idField) ||
                m.id ||
                m.name ||
                m.model_id ||
                m.model);
            if (parseConfig.idPrefix && id && id.startsWith(parseConfig.idPrefix)) {
                id = id.substring(parseConfig.idPrefix.length);
            }
            const nameField = parseConfig.nameField || parseConfig.idField || 'id';
            const name = (this.#getNestedValue(m, nameField) ||
                m.name ||
                m.display_name ||
                m.displayName ||
                id);
            let displayName = null;
            if (parseConfig.displayNameField) {
                displayName = (this.#getNestedValue(m, parseConfig.displayNameField) ||
                    m.display_name ||
                    m.displayName);
            }
            let contextWindow;
            if (parseConfig.contextWindowField) {
                contextWindow = (this.#getNestedValue(m, parseConfig.contextWindowField) ||
                    m.inputTokenLimit ||
                    m.contextWindow ||
                    m.max_tokens ||
                    m.context_window);
            }
            else {
                contextWindow = (m.inputTokenLimit ||
                    m.contextWindow ||
                    m.max_tokens ||
                    m.context_window ||
                    undefined);
            }
            return {
                id: id || 'unknown',
                provider: providerName,
                name: displayName || name || id || 'unknown',
                contextWindow: contextWindow || undefined,
                raw: model
            };
        }).filter((model) => model !== null);
    }
    static #getNestedValue(obj, path) {
        if (!path || !obj)
            return null;
        const parts = path.split('.');
        let current = obj;
        for (const part of parts) {
            if (current && typeof current === 'object' && part in current) {
                current = current[part];
            }
            else {
                return null;
            }
        }
        return current;
    }
    static clearCache(providerName = null) {
        if (providerName) {
            providerName = this.#normalizeProviderName(providerName);
            this.#modelsCache.delete(providerName);
        }
        else {
            this.#modelsCache.clear();
        }
    }
    static isCached(providerName) {
        providerName = this.#normalizeProviderName(providerName);
        return this.#modelsCache.has(providerName) && this.#modelsCache.get(providerName).size > 0;
    }
    static getCacheTimestamp(providerName) {
        providerName = this.#normalizeProviderName(providerName);
        const cached = this.#modelsCache.get(providerName);
        if (cached && cached.size > 0) {
            return Date.now();
        }
        return null;
    }
    static getDefaultModels() {
        this.init();
        const providers = this.list();
        const models = {};
        providers.forEach(provider => {
            if (provider.defaultModel) {
                models[provider.name] = provider.defaultModel;
            }
        });
        return models;
    }
    static reset() {
        this.#providers.clear();
        this.#modelsCache.clear();
        this.#apiKeys.clear();
        this.#initialized = false;
        this.init();
    }
}
export default ProviderRegistry;
