// resilient-llm/dist/lib/ResilientLLM.js
/**
 * A common interface to interact with AI models (with resilience: rate limiting, circuit breaker, retries).
 *
 * @example
 * import { ResilientLLM } from 'resilient-llm';
 * const llm = new ResilientLLM({ aiService: "anthropic", model: "claude-haiku-4-5-20251001", maxTokens: 2048, temperature: 0 });
 * const response = await llm.chat([{ role: "user", content: "Hello, world!" }]);
 * console.log(response.content); // assistant output
 * // response.metadata is always included for observability
 * // Cancel all llm operations for this instance:
 * llm.abort();
 */
import { Tiktoken } from "js-tiktoken/lite";
import o200k_base from "js-tiktoken/ranks/o200k_base";
import { randomUUID } from "node:crypto";
import ResilientOperation from "./ResilientOperation.js";
import ProviderRegistry from "./ProviderRegistry.js";
import { normalizeStructuredOutputConfig, mapConfigToRequestFields, parseStructuredResponse, validateStructuredResponse, } from "./StructuredOutput.js";
import { ResilientLLMError } from "./ResilientLLMError.js";
/**
 * ResilientLLM: unified chat interface with configurable provider, model, rate limits, circuit breaker, and retries.
 * Constructor options: aiService, model, temperature, maxTokens, timeout, cacheStore, maxInputTokens, topP,
 * rateLimitConfig, retries, backoffFactor, circuitBreakerConfig, maxConcurrent, onRateLimitUpdate.
 */
class ResilientLLM {
    static encoder;
    static _safeHeaderPrefixes = ['x-ratelimit', 'rate-limit', 'retry-after', 'x-request-id', 'request-id'];
    aiService;
    model;
    temperature;
    maxTokens;
    cacheStore;
    maxInputTokens;
    topP;
    maxCompletionTokens;
    reasoningEffort;
    retries;
    backoffFactor;
    timeout;
    rateLimitConfig;
    circuitBreakerConfig;
    maxConcurrent;
    onRateLimitUpdate;
    resilientOperations;
    llmOutOfService;
    responseFormat;
    output_config;
    _abortController;
    constructor(options) {
        this.aiService = options?.aiService || process.env.PREFERRED_AI_SERVICE || "anthropic";
        this.model = options?.model || process.env.PREFERRED_AI_MODEL || "claude-haiku-4-5-20251001";
        this.temperature = options?.temperature ?? process.env.AI_TEMPERATURE;
        this.maxTokens = options?.maxTokens || process.env.MAX_TOKENS;
        this.cacheStore = options?.cacheStore || {};
        // Default to 100k to avoid accidental context window overflow
        this.maxInputTokens = options?.maxInputTokens || process.env.MAX_INPUT_TOKENS || 100000;
        this.topP = options?.topP ?? process.env.AI_TOP_P;
        this.maxCompletionTokens = options?.maxCompletionTokens ?? process.env.MAX_COMPLETION_TOKENS;
        this.reasoningEffort = options?.reasoningEffort ?? process.env.AI_REASONING_EFFORT;
        this.retries = options?.retries || 3;
        this.backoffFactor = options?.backoffFactor || 2;
        this.timeout = options?.timeout || process.env.LLM_TIMEOUT || 60000;
        this.rateLimitConfig = options?.rateLimitConfig || { requestsPerMinute: 10, llmTokensPerMinute: 150000 };
        this.circuitBreakerConfig = options?.circuitBreakerConfig
            ? { failureThreshold: options.circuitBreakerConfig.failureThreshold ?? 5, cooldownPeriod: options.circuitBreakerConfig.cooldownPeriod ?? 30000 }
            : { failureThreshold: 5, cooldownPeriod: 30000 };
        this.maxConcurrent = options?.maxConcurrent;
        this.onRateLimitUpdate = options?.onRateLimitUpdate;
        this.responseFormat = options?.responseFormat;
        this.output_config = options?.output_config;
        this._abortController = null;
        this.resilientOperations = {}; // Store resilient operation instances for observability
    }
    /**
     * Chat with the LLM.
     * @param conversationHistory - Array of messages (role + content)
     * @param llmOptions - Overrides for this call (model, temperature, tools, apiKey, etc.)
     * @param observabilityOptions - Observability/metadata options
     * @returns A response envelope: `{ content, toolCalls?, metadata }`
     * @throws {ResilientLLMError} User-actionable failures with a stable `code` for branching,
     *   `metadata` mirroring success metadata, and `cause` for logging. See `error.code` catalog
     *   in `ResilientLLMError.ts`.
     */
    async chat(conversationHistory, llmOptions, observabilityOptions) {
        llmOptions = llmOptions ?? {};
        observabilityOptions = observabilityOptions ?? {};
        const startTime = Date.now();
        const requestId = randomUUID();
        let metadata = null;
        let resilientOperation = null;
        try {
            const preparedRequest = this._buildRequest({
                conversationHistory,
                llmOptions,
            });
            resilientOperation = new ResilientOperation({
                bucketId: preparedRequest.aiService,
                ...preparedRequest.resilienceConfig,
                collectMetrics: true,
                onRateLimitUpdate: this.onRateLimitUpdate,
                cacheStore: this.cacheStore
            });
            metadata = this._initMetadata({
                requestId,
                startTime,
                preparedRequest,
                llmOptions,
                operationId: resilientOperation.id,
            });
            observabilityOptions = { ...observabilityOptions, metadata };
            if (!this._abortController || this._abortController.signal.aborted) {
                this._abortController = new AbortController();
            }
            this.resilientOperations[resilientOperation.id] = resilientOperation;
            const { data, statusCode } = await resilientOperation
                .withTokens(preparedRequest.estimatedTokens)
                .withCache(preparedRequest.enableCache)
                .withAbortControl(this._abortController)
                .execute(this._makeHttpRequest, preparedRequest.apiUrl, preparedRequest.requestBody, preparedRequest.headers, this._abortController.signal, observabilityOptions);
            console.log("LLM chat status code:", statusCode, data?.error
                ? data.error?.message
                : undefined);
            if ([429, 529].includes(statusCode)) {
                delete this.resilientOperations[resilientOperation.id];
                return await this.retryChatWithAlternateService(conversationHistory, llmOptions);
            }
            const dataAsArray = data;
            if (data?.error || (Array.isArray(data) && dataAsArray[0]?.error)) {
                throw new Error(data?.error?.message ||
                    dataAsArray?.[0]?.error?.message);
            }
            if (statusCode !== 200) {
                // Future: handle other status codes
            }
            const parsedResponse = this._handleResponse({
                rawData: data,
                statusCode,
                chatConfig: preparedRequest.chatConfig,
                structuredOutputConfig: preparedRequest.structuredOutputConfig,
                tools: llmOptions?.tools,
            });
            const content = parsedResponse.content;
            const toolCalls = parsedResponse.toolCalls;
            // Expose finish semantics to callers via metadata for observability.
            metadata.finishReason = parsedResponse.finishReason ?? null;
            const isEmpty = content == null || (typeof content === 'string' && content.trim() === '');
            if (isEmpty) {
                console.warn("Empty response from LLM");
            }
            const analysis = this._analyzeFinishReason(data);
            if (analysis) {
                console.log(analysis.summary);
                console.log(analysis.tokensInfo);
                console.log(analysis.recommendation);
            }
            delete this.resilientOperations[resilientOperation.id];
            const usageData = data?.usage && typeof data.usage === 'object'
                ? data.usage
                : {};
            metadata = this._finalizeMetadata(metadata, 'success', resilientOperation.getRuntimeMetrics(), usageData);
            const response = {
                content: content ?? null,
                ...(toolCalls !== undefined ? { toolCalls } : {}),
                metadata,
            };
            return response;
        }
        catch (error) {
            const aiService = llmOptions?.aiService || this.aiService;
            console.error(`Error calling ${aiService} API:`, error);
            if (resilientOperation) {
                delete this.resilientOperations[resilientOperation.id];
            }
            if (metadata) {
                metadata = this._finalizeMetadata(metadata, 'error', resilientOperation?.getRuntimeMetrics() ?? null);
            }
            this.parseError(null, error, metadata);
        }
    }
    /** Resolves resilience settings from instance defaults and per-call overrides. */
    _resolveResilienceConfig(llmOptions) {
        return {
            timeout: Number(llmOptions?.timeout ?? this.timeout),
            retries: llmOptions?.retries ?? this.retries,
            backoffFactor: llmOptions?.backoffFactor ?? this.backoffFactor,
            circuitBreakerConfig: {
                failureThreshold: llmOptions?.circuitBreakerConfig?.failureThreshold ?? this.circuitBreakerConfig?.failureThreshold,
                cooldownPeriod: llmOptions?.circuitBreakerConfig?.cooldownPeriod ?? this.circuitBreakerConfig?.cooldownPeriod
            },
            rateLimitConfig: {
                requestsPerMinute: llmOptions?.rateLimitConfig?.requestsPerMinute ?? this.rateLimitConfig?.requestsPerMinute,
                llmTokensPerMinute: llmOptions?.rateLimitConfig?.llmTokensPerMinute ?? this.rateLimitConfig?.llmTokensPerMinute
            },
            maxConcurrent: llmOptions?.maxConcurrent ?? this.maxConcurrent,
        };
    }
    /** Builds provider request data from explicit chat inputs and defaults. */
    _buildRequest(input) {
        const { conversationHistory, llmOptions } = input;
        const aiService = llmOptions?.aiService || this.aiService;
        const model = llmOptions?.model || this.model;
        // Get provider configuration
        const providerConfig = ProviderRegistry.get(aiService);
        if (!providerConfig) {
            const available = ProviderRegistry.list().map(p => `"${p.name}"`).join(', ');
            throw new Error(`Invalid provider specified: "${aiService}". Available: ${available}`);
        }
        const chatConfig = ProviderRegistry.getChatConfig(aiService) || {
            messageFormat: 'openai',
            responseParsePath: 'choices[0].message.content',
            toolSchemaType: 'openai'
        };
        // Get API URL from provider configuration
        let apiUrl = ProviderRegistry.getChatApiUrl(aiService);
        if (!apiUrl) {
            const available = ProviderRegistry.list().map(p => `"${p.name}"`).join(', ');
            throw new Error(`Invalid AI service: "${aiService}". Available: ${available}`);
        }
        // Validate API key for providers that require it (check llmOptions and registry)
        const hasApiKeyInOptions = !!llmOptions?.apiKey;
        const hasApiKeyInRegistry = ProviderRegistry.hasApiKey(aiService);
        if (!hasApiKeyInOptions && !hasApiKeyInRegistry && !providerConfig.authConfig?.optional) {
            const envVars = providerConfig.envVarNames?.join(' or ') || 'API_KEY';
            throw new Error(`${envVars} is not set for provider "${aiService}"`);
        }
        // Get API key early for URL building and headers
        const apiKey = llmOptions?.apiKey || null;
        // Handle query-parameter auth; buildApiUrl uses endpoint-specific auth when needed
        apiUrl = ProviderRegistry.buildApiUrl(aiService, apiUrl, apiKey);
        const maxInputTokens = Number(llmOptions?.maxInputTokens || this.maxInputTokens);
        // Estimate LLM tokens for this request
        const estimatedTokens = ResilientLLM.estimateTokens(conversationHistory?.map(message => message?.content)?.join("\n"));
        console.log("Estimated LLM input tokens:", estimatedTokens, "/", maxInputTokens);
        if (estimatedTokens > maxInputTokens) {
            throw new Error("Input tokens exceed the maximum limit of " + maxInputTokens);
        }
        const requestBody = { model };
        if (llmOptions?.tools) {
            requestBody.tools = llmOptions.tools;
        }
        const structuredOutputConfigResult = normalizeStructuredOutputConfig(llmOptions?.responseFormat ?? this.responseFormat, llmOptions?.output_config ?? this.output_config);
        if (!structuredOutputConfigResult.ok) {
            const soErr = structuredOutputConfigResult.error;
            throw new ResilientLLMError(soErr.message, soErr.code, {
                cause: soErr,
            });
        }
        const normalizedStructuredOutputConfig = structuredOutputConfigResult.data;
        const structuredOutputConfig = normalizedStructuredOutputConfig._source === 'none'
            ? null
            : normalizedStructuredOutputConfig;
        // Map structured output config to provider request fields
        const structuredRequestFields = mapConfigToRequestFields(normalizedStructuredOutputConfig, chatConfig.structuredOutputRequestField ?? 'response_format');
        if (structuredRequestFields.response_format !== undefined) {
            requestBody.response_format = structuredRequestFields.response_format;
        }
        if (structuredRequestFields.output_config !== undefined) {
            requestBody.output_config = structuredRequestFields.output_config;
        }
        if (model?.startsWith("o") || model?.startsWith("gpt-5")) {
            // Reasoning model parameters
            const maxCompletionTokens = llmOptions?.maxCompletionTokens
                ?? this.maxCompletionTokens
                ?? llmOptions?.maxTokens
                ?? this.maxTokens;
            if (maxCompletionTokens != null) {
                requestBody.max_completion_tokens = Number(maxCompletionTokens);
            }
            const reasoningEffort = llmOptions?.reasoningEffort ?? this.reasoningEffort;
            if (reasoningEffort != null) {
                requestBody.reasoning_effort = reasoningEffort;
            }
        }
        else {
            requestBody.max_tokens = Number(llmOptions?.maxTokens || this.maxTokens);
            const temperature = llmOptions?.temperature ?? this.temperature;
            if (temperature != null) {
                requestBody.temperature = Number(temperature);
            }
            const topP = llmOptions?.topP ?? this.topP;
            if (topP != null) {
                requestBody.top_p = Number(topP);
            }
        }
        // Format messages based on provider configuration
        if (chatConfig.messageFormat === 'anthropic') {
            const { system, messages } = this.formatMessageForAnthropic(conversationHistory);
            if (system)
                requestBody.system = system;
            requestBody.messages = messages;
        }
        else {
            // Default: 'openai' format (keep system in messages)
            requestBody.messages = conversationHistory;
            if (process.env.STORE_AI_API_CALLS === 'true' && providerConfig.name === 'openai') {
                requestBody.store = true;
            }
        }
        // Handle tool schema conversion based on provider
        if (requestBody.tools?.length) {
            const toolDefinitions = JSON.parse(JSON.stringify(requestBody.tools));
            if (chatConfig.toolSchemaType === 'anthropic') {
                // Convert to Anthropic format (input_schema)
                for (const tool of toolDefinitions) {
                    if (!tool.function.input_schema && tool.function.parameters) {
                        tool.function.input_schema = tool.function.parameters;
                        delete tool.function.parameters;
                    }
                }
            }
            else {
                // Convert to OpenAI format (parameters)
                for (const tool of toolDefinitions) {
                    if (!tool.function.parameters && tool.function.input_schema) {
                        tool.function.parameters = tool.function.input_schema;
                        delete tool.function.input_schema;
                    }
                }
            }
            requestBody.tools = toolDefinitions;
        }
        const defaultHeaders = {
            'Content-Type': 'application/json'
        };
        const headers = ProviderRegistry.buildAuthHeaders(aiService, apiKey, defaultHeaders, apiUrl);
        return {
            apiUrl,
            requestBody,
            headers,
            chatConfig,
            structuredOutputConfig,
            estimatedTokens,
            aiService,
            model,
            maxInputTokens,
            resilienceConfig: this._resolveResilienceConfig(llmOptions),
            enableCache: llmOptions?.enableCache ?? true,
        };
    }
    /** Parses and normalizes the provider response into caller-facing content.
     * @param input - The input object containing the raw data, chat configuration, and tools.
     * @returns The parsed chat response.
     * @throws {ResilientLLMError} If the structured output parsing or validation fails.
     * @example
     * const parsedChatResponse = this._handleResponse({
     *   rawData: rawProviderResponse,
     *   chatConfig: chatConfig,
     *   tools: tools,
     *   structuredOutputConfig: structuredOutputConfig,
     * });
     * // {
     * //   content: 'The capital of France is Paris.',
     * //   finishReason: 'stop',
     * // }
    */
    _handleResponse(input) {
        const envelope = this.parseChatCompletion(input.rawData, input.chatConfig, input.tools);
        let content = envelope.content ?? null;
        let toolCalls;
        if (envelope.toolCalls) {
            // In tool mode, keep assistant payload + tool calls separate.
            toolCalls = envelope.toolCalls;
        }
        else if (input.structuredOutputConfig && input.structuredOutputConfig.expectsJson) {
            const parseResult = parseStructuredResponse(envelope, input.structuredOutputConfig);
            if (!parseResult.ok) {
                console.error("[ResilientLLM][_handleResponse] Structured output parsing failed", {
                    structuredOutputConfig: input.structuredOutputConfig,
                    envelopeContent: envelope.content,
                    // Raw provider payload to debug response-shape mismatches.
                    rawProviderResponse: input.rawData,
                    rawResponseForError: parseResult.error?.rawResponse ?? undefined,
                    structuredOutputError: {
                        code: parseResult.error?.code ?? undefined,
                        message: parseResult.error.message,
                        validation: parseResult.error?.validation ?? null,
                    }
                });
                const soErr = parseResult.error;
                throw new ResilientLLMError(soErr.message, soErr.code, {
                    cause: soErr,
                });
            }
            if (parseResult.data &&
                typeof parseResult.data === 'object' &&
                !(Array.isArray(parseResult.data))) {
                const validateResult = validateStructuredResponse(parseResult.data, input.structuredOutputConfig);
                if (!validateResult.ok) {
                    console.error("[ResilientLLM][_handleResponse] Structured output validation failed", {
                        structuredOutputConfig: input.structuredOutputConfig,
                        parsedJson: parseResult.data,
                        envelopeContent: envelope.content,
                        rawProviderResponse: input.rawData,
                        validationError: validateResult.error,
                    });
                    const soErr = validateResult.error;
                    throw new ResilientLLMError(soErr.message, soErr.code, {
                        cause: soErr,
                    });
                }
                content = validateResult.data;
            }
            else {
                content = parseResult.data;
            }
        }
        return {
            content,
            ...(toolCalls !== undefined ? { toolCalls } : {}),
            finishReason: envelope.finishReason ?? null,
        };
    }
    /** Initializes operation metadata before request execution starts. */
    _initMetadata(params) {
        const { requestId, startTime, preparedRequest, llmOptions, operationId } = params;
        const temperature = llmOptions?.temperature ?? this.temperature;
        const topP = llmOptions?.topP ?? this.topP;
        const effectiveMaxTokens = (preparedRequest.requestBody.max_completion_tokens
            ?? preparedRequest.requestBody.max_tokens
            ?? null);
        return {
            requestId,
            operationId,
            startTime,
            finishReason: null,
            config: {
                aiService: preparedRequest.aiService,
                model: preparedRequest.model,
                temperature: temperature != null ? Number(temperature) : null,
                maxTokens: effectiveMaxTokens,
                topP: topP != null ? Number(topP) : null,
                maxInputTokens: preparedRequest.maxInputTokens,
                estimatedInputTokens: preparedRequest.estimatedTokens,
                enableCache: preparedRequest.enableCache,
                ...preparedRequest.resilienceConfig,
            },
            events: [],
            timing: { totalTimeMs: null, rateLimitWaitMs: 0, httpRequestMs: null },
            retries: [],
            rateLimiting: { requestedTokens: preparedRequest.estimatedTokens, totalWaitMs: 0 },
            circuitBreaker: {},
            http: {},
            cache: { enabled: preparedRequest.enableCache },
            service: { attempted: [preparedRequest.aiService], final: preparedRequest.aiService },
        };
    }
    /** Finalizes metadata by merging runtime metrics and computing total timing. */
    _finalizeMetadata(base, phase, runtimeMetrics, usageData) {
        const prev = base ?? {};
        const timingPrev = prev.timing ?? {};
        const totalTimeMs = typeof prev.startTime === 'number'
            ? Date.now() - prev.startTime
            : timingPrev.totalTimeMs ?? null;
        const timing = {
            totalTimeMs,
            rateLimitWaitMs: runtimeMetrics?.rateLimiting?.totalWaitMs ?? timingPrev.rateLimitWaitMs ?? 0,
            httpRequestMs: timingPrev.httpRequestMs ?? null,
        };
        const rateLimiting = {
            requestedTokens: prev.rateLimiting?.requestedTokens ?? runtimeMetrics?.rateLimiting?.requestedTokens ?? 0,
            totalWaitMs: runtimeMetrics?.rateLimiting?.totalWaitMs
                ?? prev.rateLimiting?.totalWaitMs
                ?? 0,
        };
        const circuitBreaker = {
            ...(prev.circuitBreaker ?? {}),
            ...(runtimeMetrics?.circuitBreaker ?? {}),
        };
        const cache = {
            ...(prev.cache ?? {}),
            ...(runtimeMetrics?.cache ?? {}),
        };
        const usage = phase === 'success' && usageData && typeof usageData === 'object'
            ? (() => {
                const pt = usageData.prompt_tokens;
                const ct = usageData.completion_tokens;
                const tot = usageData.total_tokens ??
                    (pt != null && ct != null ? pt + ct : null);
                return {
                    prompt_tokens: pt ?? null,
                    completion_tokens: ct ?? null,
                    total_tokens: tot ?? null,
                };
            })()
            : prev.usage;
        const retries = runtimeMetrics?.retries ?? prev.retries ?? [];
        return {
            ...prev,
            timing,
            retries,
            rateLimiting,
            circuitBreaker,
            cache,
            usage,
            service: prev.service ?? { attempted: [this.aiService], final: this.aiService },
        };
    }
    async retryChatWithAlternateService(conversationHistory, llmOptions) {
        llmOptions = llmOptions ?? {};
        this.llmOutOfService = this.llmOutOfService || [];
        const currentService = llmOptions?.aiService || this.aiService;
        this.llmOutOfService.push(currentService);
        const defaultModels = ProviderRegistry.getDefaultModels();
        for (const [providerName, defaultModel] of Object.entries(defaultModels)) {
            if (!this.llmOutOfService.includes(providerName)) {
                console.log("Switching LLM service to:", providerName, defaultModel);
                const newLLMOptions = Object.assign(llmOptions, {
                    aiService: providerName,
                    model: defaultModel
                });
                return this.chat(conversationHistory, newLLMOptions);
            }
        }
        throw new Error("No alternative model found");
    }
    async _makeHttpRequest(apiUrl, requestBody, headers, abortSignal, observabilityOptions) {
        console.log("Making HTTP request to:", apiUrl);
        console.log("You may cancel it by calling abort() method on the ResilientLLM instance");
        const httpStartTime = Date.now();
        try {
            const response = await fetch(apiUrl, {
                method: 'POST',
                headers: headers,
                body: JSON.stringify(requestBody),
                signal: abortSignal
            });
            console.log("Response Headers:", response?.headers);
            const data = await response?.json();
            const result = { data, statusCode: response?.status };
            const httpDurationMs = Date.now() - httpStartTime;
            console.log(`Request to ${apiUrl} completed in ${httpDurationMs} ms`);
            if (observabilityOptions?.metadata) {
                ResilientLLM._captureHttpMetadata(observabilityOptions.metadata, apiUrl, response, httpDurationMs);
            }
            return result;
        }
        catch (error) {
            const httpDurationMs = Date.now() - httpStartTime;
            console.log(`Request to ${apiUrl} failed in ${httpDurationMs} ms`);
            if (observabilityOptions?.metadata) {
                ResilientLLM._captureHttpMetadata(observabilityOptions.metadata, apiUrl, null, httpDurationMs, error);
            }
            if (error.name === 'AbortError') {
                console.error(`Request to ${apiUrl} timed out or was cancelled`);
            }
            console.error(`Error in request to ${apiUrl}:`, error);
            throw error;
        }
    }
    static _captureHttpMetadata(metadata, apiUrl, response, durationMs, error) {
        let sanitizedUrl;
        try {
            const parsed = new URL(apiUrl);
            parsed.search = '';
            sanitizedUrl = parsed.toString();
        }
        catch {
            sanitizedUrl = apiUrl;
        }
        const safeHeaders = {};
        if (response?.headers) {
            for (const [name, value] of response.headers.entries()) {
                const lower = name.toLowerCase();
                if (ResilientLLM._safeHeaderPrefixes.some(prefix => lower.startsWith(prefix) || lower === prefix)) {
                    safeHeaders[name] = value;
                }
            }
        }
        metadata.http = {
            url: sanitizedUrl,
            method: 'POST',
            statusCode: response?.status ?? null,
            headers: safeHeaders,
            durationMs,
            ...(error ? { error: error.message } : {}),
        };
        if (!metadata.timing) {
            metadata.timing = { totalTimeMs: null, rateLimitWaitMs: 0, httpRequestMs: null };
        }
        metadata.timing.httpRequestMs = durationMs;
    }
    parseError(statusCode, error, operationMetadata) {
        if (error instanceof ResilientLLMError) {
            if (operationMetadata && !error.metadata) {
                throw new ResilientLLMError(error.message, error.code, {
                    cause: error.cause,
                    metadata: operationMetadata,
                    retryable: error.retryable,
                });
            }
            throw error;
        }
        const { message, code } = ResilientLLM._mapHttpStatus(statusCode, error);
        const metadata = {
            ...(operationMetadata ?? {}),
            provider: { httpStatus: statusCode },
        };
        throw new ResilientLLMError(message, code, {
            cause: error,
            metadata,
        });
    }
    /** Maps an HTTP status to a stable error code and message. */
    static _mapHttpStatus(statusCode, error) {
        switch (statusCode) {
            case 400:
                console.error("Bad request");
                return { message: error?.message || "Bad request", code: "PROVIDER_BAD_REQUEST" };
            case 401:
                console.error("Invalid API Key");
                return { message: error?.message || "Invalid API Key", code: "PROVIDER_UNAUTHORIZED" };
            case 403:
                return { message: error?.message || "You are not authorized to access this resource", code: "PROVIDER_FORBIDDEN" };
            case 429:
                return { message: error?.message || "Rate limit exceeded", code: "PROVIDER_RATE_LIMIT" };
            case 404:
                return { message: error?.message || "Not found", code: "PROVIDER_NOT_FOUND" };
            case 500:
                return { message: error?.message || "Internal server error", code: "PROVIDER_INTERNAL_ERROR" };
            case 503:
                return { message: error?.message || "Service unavailable", code: "PROVIDER_UNAVAILABLE" };
            case 529:
                return { message: error?.message || "API temporarily overloaded", code: "PROVIDER_OVERLOADED" };
            default:
                return { message: error?.message || "Unknown error", code: "PROVIDER_ERROR" };
        }
    }
    /**
     * Converts the messages array to the format required by Anthropic
     * Extracts system messages into a separate system field
     * @param {Array} messages - Array of message objects
     * @returns {Object} Object with system (string) and messages (array) properties
     * @example
     * const llm = new ResilientLLM({ aiService: "anthropic", model: "claude-haiku-4-5-20251001" });
     * const { system, messages } = llm.formatMessageForAnthropic(originalMessages);
     * // originalMessages: [{ role: "system", content: "You are a helpful assistant." }, { role: "user", content: "Hello, world!" }]
     * // system: "You are a helpful assistant."
     * // messages: [{ role: "user", content: "Hello, world!" }]
     */
    formatMessageForAnthropic(messages) {
        let system;
        const messagesWithoutSystemMessage = [];
        for (let i = 0; i < messages.length; i++) {
            if (messages[i].role === "system" && messages[i].content) {
                system = messages[i].content;
            }
            else {
                messagesWithoutSystemMessage.push(messages[i]);
            }
        }
        return { system, messages: messagesWithoutSystemMessage };
    }
    /**
     * Analyze and summarize finish_reason and token usage, with actionable plain-English recommendations.
     * @param {Object} data - Raw API response data
     * @param {number} data.choices[0].finish_reason - Finish reason
     * @param {Object} data.usage - Usage object
     * @param {number} data.usage.prompt_tokens - Prompt tokens
     * @param {number} data.usage.completion_tokens - Completion tokens
     * @param {number} data.usage.total_tokens - Total tokens
     * @returns {Object} Object with summary, tokensInfo, and recommendation properties
     * @example
     * const { summary, tokensInfo, recommendation } = llm._analyzeFinishReason(data);
     * // data: { choices: [{ finish_reason: "stop" }], usage: { prompt_tokens: 10, completion_tokens: 10, total_tokens: 20 } }
     * // summary: "✅ The model replied fully and finished normally with no issues."
     * // tokensInfo: "Prompt tokens used: 10 | Completion tokens used: 10 | Total tokens: 20"
     * // recommendation: "Tip: You used 10 prompt tokens and 10 for the reply (total: 20). For best performance and efficiency, try to keep the total under 1500 tokens."
     * @private
     */
    _analyzeFinishReason(data) {
        if (!data)
            return null;
        const choices = data?.choices;
        const choice = choices?.[0];
        const finishReason = choice?.finish_reason ?? null;
        const usage = (data?.usage || {});
        const hasPromptTokens = usage?.prompt_tokens != null;
        const hasCompletionTokens = usage?.completion_tokens != null;
        const promptTokens = hasPromptTokens ? usage.prompt_tokens : null;
        const completionTokens = hasCompletionTokens ? usage.completion_tokens : null;
        const totalTokens = usage?.total_tokens ??
            ((hasPromptTokens && hasCompletionTokens)
                ? promptTokens + completionTokens
                : null);
        let summary = "";
        let recommendation = "";
        let tokensInfo = "";
        const calculatedRecommendedMax = (hasCompletionTokens && completionTokens !== null)
            ? Math.ceil(completionTokens * 1.2)
            : undefined;
        switch (finishReason) {
            case "length":
                summary =
                    "⚠️ The model stopped because it reached the set output token limit and couldn't finish its answer.";
                recommendation =
                    calculatedRecommendedMax
                        ? `Recommendation: To get complete answers, set max_tokens (or max_completion_tokens) to at least ${calculatedRecommendedMax} (20% above what was just used), or trim your input prompt/context.`
                        : "Recommendation: To get complete answers, increase max_tokens (or max_completion_tokens) or trim your input prompt/context.";
                break;
            case "content_filter":
                summary =
                    "⚠️ The model stopped due to violating content policy or safety filters.";
                recommendation =
                    "Recommendation: Review your prompt and output for sensitive or restricted content, and revise as needed.";
                break;
            case "function_call":
                summary =
                    "⛏️ The model paused to perform a function or tool call as requested.";
                recommendation =
                    "Recommendation: Handle the tool or function call with your code, then continue the conversation.";
                break;
            case "stop":
                summary =
                    "✅ The model replied fully and finished normally with no issues.";
                if ((totalTokens != null && totalTokens > 0) && (hasPromptTokens || hasCompletionTokens)) {
                    const tokenParts = [];
                    if (hasPromptTokens)
                        tokenParts.push(`${promptTokens} prompt tokens`);
                    if (hasCompletionTokens)
                        tokenParts.push(`${completionTokens} for the reply`);
                    recommendation =
                        `Tip: You used${tokenParts.length ? ' ' + tokenParts.join(' and ') : ''}${totalTokens !== null ? ` (total: ${totalTokens})` : ""}. For best performance and efficiency, try to keep the total under ${Math.max(1500, totalTokens ?? 1500)} tokens.`;
                }
                break;
            case null:
            case undefined:
                summary = "ℹ️ No finish reason reported.";
                break;
            default:
                summary = `ℹ️ Response stopped with finish_reason: "${finishReason}".`;
                recommendation = "Recommendation: Check the API documentation for this finish_reason.";
        }
        if (hasPromptTokens || hasCompletionTokens || totalTokens != null) {
            tokensInfo = [
                hasPromptTokens
                    ? `Prompt tokens used: ${promptTokens}`
                    : "",
                hasCompletionTokens
                    ? `Completion tokens used: ${completionTokens}`
                    : "",
                totalTokens != null
                    ? `Total tokens: ${totalTokens}`
                    : ""
            ]
                .filter(Boolean)
                .join(" | ");
        }
        else {
            tokensInfo = "";
        }
        return { summary, tokensInfo, recommendation };
    }
    /**
     * Extracts provider response into a normalized envelope for downstream processing.
     * Responsible for content extraction and tool-call detection only — no schema logic.
     */
    parseChatCompletion(data, chatConfig, tools) {
        if (!data) {
            return { content: null };
        }
        const parsePath = chatConfig?.responseParsePath || 'choices[0].message.content';
        const content = this._getNestedValue(data, parsePath);
        const choices = data?.choices;
        const finishReason = choices?.[0]?.finish_reason ?? null;
        if (tools) {
            const dataContent = data?.content;
            const toolCalls = choices?.[0]?.message
                ? choices[0].message?.tool_calls
                : dataContent?.[0]?.tool_use || null;
            if (toolCalls) {
                return { content, toolCalls, finishReason };
            }
        }
        return { content, finishReason };
    }
    /**
     * Helper method to get nested value from object using path notation
     * Supports both dot notation and bracket notation (e.g., 'choices[0].message.content')
     * @private
     * @example
     * const value = _getNestedValue(data, 'choices[0].message.content');
     * // data: { choices: [{ message: { content: "Hello, world!" } }] }
     * // value: "Hello, world!"
     */
    _getNestedValue(obj, path) {
        if (!path || !obj)
            return null;
        const parts = path.split(/[.\[\]]+/).filter(p => p);
        let current = obj;
        for (const part of parts) {
            if (current == null) {
                return null;
            }
            const index = parseInt(part, 10);
            if (!isNaN(index) && Array.isArray(current) && index >= 0 && index < current.length) {
                current = current[index];
            }
            else if (current && typeof current === 'object' && part in current) {
                current = current[part];
            }
            else {
                return null;
            }
        }
        return current;
    }
    /** @deprecated Use parseChatCompletion with chatConfig instead */
    parseOpenAIChatCompletion(data, tools) {
        if (tools) {
            const choices = data?.choices;
            const message = choices?.[0]?.message;
            return { content: message?.content, toolCalls: message?.tool_calls };
        }
        const choices = data?.choices;
        const message = choices?.[0]?.message;
        return message?.content;
    }
    /** @deprecated Use parseChatCompletion with chatConfig instead */
    parseAnthropicChatCompletion(data, _tools) {
        const content = data?.content;
        return content?.[0]?.text;
    }
    /** @deprecated Use parseChatCompletion with chatConfig instead */
    parseOllamaChatCompletion(data, _tools) {
        return data?.response;
    }
    /** @deprecated Use parseChatCompletion with chatConfig instead */
    parseGoogleChatCompletion(data, _tools) {
        const choices = data?.choices;
        const message = choices?.[0]?.message;
        return message?.content;
    }
    /** Cancel all in-flight operations for this instance. */
    abort() {
        this._abortController?.abort();
        this._abortController = null;
        this.resilientOperations = {};
    }
    /** Estimate token count for text (uses tiktoken when available; fallback ~chars/4). */
    static estimateTokens(text) {
        if (text.length > 10000) {
            return Math.ceil(text.length / 4);
        }
        if (!ResilientLLM.encoder) {
            ResilientLLM.encoder = new Tiktoken(o200k_base);
        }
        return ResilientLLM.encoder.encode(text).length;
    }
}
export default ResilientLLM;
